src/hotspot/share/nmt/memReporter.cpp

```diff
@@ -898,4 +898,17 @@ void MemDetailDiffReporter::diff_virtual_memory_site(const NativeCallStack* stac
   }
 
   out->print_cr(")\n");
- }
+}
+
+void MemDetailReporter::report_physical_devices() {
+  MemoryFileTracker::Instance::Locker lock;
+  const GrowableArrayCHeap<MemoryFileTracker::MemoryFile*, mtNMT>& devices =
+      MemoryFileTracker::Instance::devices();
+  this->output()->cr();
+  this->output()->print_cr("Memory file details");
+  this->output()->cr();
+  for (int i = 0; i < devices.length(); i++) {
+    MemoryFileTracker::MemoryFile* dev = devices.at(i);
+    MemoryFileTracker::Instance::print_report_on(dev, this->output(), scale());
+  }
+}
```

Comment:
Does `devices.length()` and `devices.at(i)` are really needed to be exposed?

Can we consider pushing all this inside `MemoryFileTracker`?

We make 4 different API calls to MemoryFileTracker here in such a small function.

Reply:
Sure, I prefer having all of the respective reporting code close to the classes instead of in the `memReporter.cpp` file. It'd be more in the current style to move the `print_report_on` method from `MemoryFileTracker` into `MemDetailReporter` instead.